docs/contributing/mcp-apps-architecture.mdx

```diff
@@ -0,0 +1,326 @@
+
+```javascript
+// Widget sends
+{ jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: "...", arguments: {} } }
+
+// Host responds
+{ jsonrpc: "2.0", id: 1, result: { ... } }
+// or
+{ jsonrpc: "2.0", id: 1, error: { code: -32000, message: "..." } }
+```
+
+### Notifications (no response)
+
+```javascript
+// Host sends
+{ jsonrpc: "2.0", method: "ui/notifications/tool-result", params: { ... } }
+```
+
+### Supported methods
+
+| Method | Direction | Description |
+|--------|-----------|-------------|
+| `ui/initialize` | Widget → Host | Initialize widget, get host context |
+| `tools/call` | Widget → Host | Call an MCP tool |
+| `resources/read` | Widget → Host | Read an MCP resource |
+| `ui/open-link` | Widget → Host | Open external URL |
+| `ui/message` | Widget → Host | Send message to chat |
+| `ui/size-change` | Widget → Host | Notify of size change |
+| `ui/notifications/initialized` | Widget → Host | Widget finished initializing |
+| `ui/notifications/tool-input` | Host → Widget | Send tool input |
+| `ui/notifications/tool-result` | Host → Widget | Send tool result |
+| `ui/host-context-change` | Host → Widget | Theme or context changed |
+
+## Server routes
+
+### POST /api/mcp/apps/widget/store
+
+Stores widget data for later retrieval by the iframe.
+
+### GET /api/mcp/apps/widget-content/:toolId
+
+Fetches widget HTML from MCP resource and injects the client script.
+
+### GET /api/mcp/sandbox-proxy
+
+Serves the sandbox proxy HTML that creates the double-iframe architecture.
+
+## Security considerations
+
+### Content Security Policy
+
+The sandbox proxy uses a permissive CSP to allow widget content:
+
+```html
+<meta http-equiv="Content-Security-Policy" content="
+  default-src 'self';
+  script-src 'self' 'unsafe-inline' 'unsafe-eval' blob: data: https://cdn.tailwindcss.com ...;
+  style-src * blob: data: 'unsafe-inline';
+  connect-src *;
+  ...
+">
+```
+
+### Iframe sandbox attributes
+
+```html
+<iframe sandbox="allow-scripts allow-same-origin allow-forms allow-popups" />
+```
+
+### Security trade-offs
+
+- Double-iframe provides origin isolation
+- `allow-same-origin` required for localStorage access
+- Widgets should be treated as semi-trusted code
+- CSP headers restrict network access
+
+## Related files
+
+- `client/src/components/chat-v2/mcp-apps-renderer.tsx` - Main renderer component
+- `client/src/components/ui/sandboxed-iframe.tsx` - Reusable double-iframe component
+- `client/src/lib/mcp-apps-utils.ts` - Detection utilities
+- `server/routes/mcp/apps.ts` - Widget storage and serving
+- `server/routes/mcp/sandbox-proxy.html` - Sandbox proxy HTML
+- `server/routes/mcp/index.ts` - Route mounting
+
+## Comparison with OpenAI Apps SDK
+
+| Feature | MCP Apps (SEP-1865) | OpenAI Apps SDK |
+|---------|---------------------|-----------------|
+| Protocol | JSON-RPC 2.0 | Custom postMessage |
+| Sandbox | Double-iframe | Single iframe |
+| API | `window.mcpApp` | `window.openai` |
```

Comment:
`window.mcpApp` does not exist - it's not part of the MCP Apps spec or any SDK. The SEP defines only native `postMessage`s